In [74]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Lab 1 - Optical Instruments
### 1.0.1Objectives
- Further explore the index of refraction.
- Explore image formation by spherical mirrors and lenses.
- Learn how to approach multiple-optic systems.



## Experiment 3 - Simple Magnifier
### Equipments
- Optical bench
- Bench lamp
- F object
- Lens C
- Printed arrow
- Phone camera
- Ruler
<br>
### Measure
1. By holding a pencil out in front of our eyes, and slowly drawing it closer to us until we are unable to keep it in focus, we determined the distance between our near points and our eyes, __near point distance__ $d_{NP} = 9.9 \pm 0.1cm$ from the following data.

In [13]:
df = pd.read_csv("nearpointdistance.csv")
df

Trial  Danny's   Heidi's  error
0        1       9.8      9.8    0.1
1        2      10.0     10.0    0.1
2  average       9.9      9.9    NaN

### Setup
<br>
1. Angle the optical bench towards the end of the table so we can comfortably sit and look down the length of it.<br>
2. Mount Lens C in a bench stand at the closer-to-us end of the optical bench.<br>
3. Mount the printed arrow in a bench stand and place it slightly less than one focal length beyond Lens C. <br>

#### A Qualitative Procedure

- Remove Lens C and place our eyes so that the printed arrow is at our near point.
- Replace Lens C and look at the printed arrow through the lens.
<br><br>

##### Observation:

<br>
- We estimated that the __angular magnification__ of the eyepiece is about two times the original object. 
<br>
- Theoretially predicted value <br><center>$m_{eyepiece} = \dfrac{d_{NP}}{f_eyepiece} = \dfrac{9.9}{5.4} = 1.83cm$</center><br>
- Our theoretical prediction roughly agrees with our estimation.
<br><br>

#### More Quantitative Procedure

<br>
- Turn off the automatic focus setting of the camera and place it just before the lens.
- Take a photo of the magnified arrow.
- Remove Lens C and take another photo of the unmagnified arrow without moving the position of camera.
- Record the distance between the camera, lens, and printed arrow.


### Data Collection   

#### Unmagnified original arrow:   
<br><br><br><br><br>
<img style = "transform: rotate(90deg)" src="unmagnified_focused.png" width="800"/> 
<br><br><br><br><br><br>
#### Magnified arrow:
<br><br><br><br><br><br>
<img style = "transform: rotate(90deg)" src="magnified_1.png" width="800"/>
<br><br><br>
#### The distance between the camera and screen is <br><center>$d_{cs} = 9.2\pm0.1cm$</center>
<br><br><br>
### Data Analysis
<br><br>
By selecting the area of the photo that contains the arrow, we observed and compared the number of vertical pixels that was occupied and used it as the height of arrows respectively.
<br><br>
<img src = "unmagnified_size.png">
<img src = "magnified_size.png">
<br><br>
From the screenshots above, we yielded that: <br>height of the unmagnified arrow $h_{u} = 380$ pixel <br>and height of the magnified arrow $h_{m} = 637$ pixel   
<br><br><br>
Using the equation $M_{EP} \equiv \dfrac{\theta_{EP}}{\theta_{NP}} = -\dfrac{d_{NP}}{f_{EP}}$, we predicted that the angular magnification is $M_{predict} = 1.83$. <br><br>
Using the equation $M_{EP} = \dfrac{h_{m}}{h_{u}}$, we determined that the angular magnification is $M_{deter} = 1.68$.
<br><br>
Thus our determined and predicted angular magnifications are roughly in agreement as well.|


In [35]:
mep = 637/380
mep

1.6763157894736842

## Experiment 4 - A Plethora of Focal Lengths

In this experiment we are going to measure the focal lengths of Lens D (a diverging lens), Mirror E (a converging mirror), and Mirror F (a diverging mirror).<br><br>


### Mirror E


#### Method 1 - Lens Clock

__Calibration:__ 
<br>
- The dial of the clock is calibrated to display the optical power of a surface made of crown glass (n = 1.53). To read curvatures, therefore, we need to make a correction:<br>
<center>$\dfrac{1}{𝑅} = \dfrac{lensclockreading}{0.53}$</center><br>
- Our lens clock DC offset was precisely calibrated to zero.
<br><br>

__Measure__   
Two independent trials both yielded the result <br>
curvature $C = -1.5\pm0.25 unit$.
<br><br>
__Analysis__<br>
$R_{1}=\dfrac{-1.50}{0.53}= -2.83\pm0.13$<br><br>
The focal length of mirror E is <br><center>$f\approx\frac{R_{1}}{2}=-1.42\pm0.07$</center><br><br>



<img src="lensD.png"/>
<img src="mirrorF.png"/>
<br><br><br><br><br><br><br><br><br><br>




#### Method 2  

<br>

##### Procedure

1. Mount Mirror E, F object, and the bench lamp as the diagram below shows. Place the screen slightly to the side of the optical bench. <br>
<img src="mirrorE.png" width = "600"/><br><br>
2. Adjust the direction mirror E faces and acquire an image of the F object on the screen.<br>
3. Measure the height of F image $h_{f}$, distance between Mirror E and the screen $d_{i}$ as well as the distance between Mirror E and the F object $d_{o}$ and record them.<br>
4. Change the distances and collect multiple independent data.

#### Data Collection


In [163]:
pf = pd.read_csv("mirrorE.csv")
pf

Trial   Hf    Di    Do  del_Di  del_Do  Del_Hf   Ho
0      1  1.0 -26.6  38.2     0.1     0.1    0.05  1.5
1      2  0.7 -22.8  52.8     0.1     0.1    0.05  1.5
2      3  0.5 -21.1  63.6     0.1     0.1    0.05  1.5
3      4  2.1 -38.5  26.9     0.1     0.1    0.05  1.5

#### Data Analysis

##### Calculate focal length of mirror E

In [164]:
pf["1/Do"] = 1/pf["Do"]
pf["1/Di"] = 1/pf["Di"]
pf["1/f"] = pf["1/Do"] + pf["1/Di"]
pf["f"] = 1/pf["1/f"]
pf

Trial   Hf    Di    Do  del_Di  del_Do  Del_Hf   Ho      1/Do      1/Di  \
0      1  1.0 -26.6  38.2     0.1     0.1    0.05  1.5  0.026178 -0.037594   
1      2  0.7 -22.8  52.8     0.1     0.1    0.05  1.5  0.018939 -0.043860   
2      3  0.5 -21.1  63.6     0.1     0.1    0.05  1.5  0.015723 -0.047393   
3      4  2.1 -38.5  26.9     0.1     0.1    0.05  1.5  0.037175 -0.025974   

        1/f          f  
0 -0.011416 -87.596552  
1 -0.024920 -40.128000  
2 -0.031670 -31.575529  
3  0.011201  89.280172

In [165]:
f = np.mean(pf["f"])
Do = np.mean(pf["Do"])
Di = np.mean(pf["Di"])
print(f)
print(Do)
print(Di)


-17.504977180527383
45.375
-27.25


In [171]:
a = 1/(1/45.375 + 1/-27.35)**2
b = np.sqrt((0.1/-27.25)**2 + (0.1/45.375)**2)
a*b


20.291115644196054

In [172]:
#Verifying the linear magnification equation

pf["di/do"] = pf["Di"]/pf["Do"]
pf["hi/ho"] = pf["Hf"]/pf["Ho"]

Rhd = pf["hi/ho"].corr(pf["di/do"])
Rhd


-0.9988717992418155

### error propagation 
$\delta_{f} = \sqrt{\left(\dfrac{\dfrac{\delta_{di}}{d_{i}^2}}{\left(\dfrac{1}{d_{o}} + \dfrac{1}{d_{i}}\right)^{2}}\right)^{2} + \left(\dfrac{\dfrac{\delta_{do}}{d_{o}^2}}{\left(\dfrac{1}{d_{o}} + \dfrac{1}{d_{i}}\right)^{2}}\right)^{2}} = 20.3 $

Magnification Equation:<br>
$M\equiv\dfrac{h_{i}}{h_{o}}=-\dfrac{d_{i}}{d_{o}}$

### (b) Diverging Lens D

#### Method 1 - Lens Clock

__Calibration:__ 
<br>
- The dial of the clock is calibrated to display the optical power of a surface made of crown glass (n = 1.53). To read curvatures, therefore, we need to make a correction:<br>
<center>$\dfrac{1}{𝑅} = \dfrac{lensclockreading}{0.53}$</center><br>
- Our lens clock DC offset was precisely calibrated to zero.
<br><br>

__Measure__   
Two independent trials both yielded the result <br>
curvature $C_{front} = C_{back} = -3.0\pm0.25 unit$.
<br><br>
__Analysis__<br>
$R_{1}=\dfrac{-3.0}{0.53}= -5.66\pm0.13$<br><br>
$R_{2}=-\dfrac{-3.0}{0.53}=5.66\pm0.13$<br><br>
$n_{lensE} = 1.53; n_{env} = 1$<br><br>
Using the lensmaker's equation:<br>
<center>$\dfrac{1}{f_{lens}} = \left(\dfrac{n_{lens}}{n_{env}}-1\right)\left(\dfrac{1}{R_{1}}-\dfrac{1}{R_{2}}\right)$</center><br>
The focal length of Lens D is <br>$f_{D} = \dfrac{1}{(1.53 - 1)(\frac{1}{-5.66}-\frac{1}{5.66})}=-5.34\pm0.017cm$


#### Method 2

##### Procedure

1. Mount lens D, lens A, F object, the screen, and the bench lamp as the diagram below shows. <br>
<img src="lensD.png" width = "600"/><br><br>
2. Adjust the height and direction lenses are facing so that they all aligh with the optical axis.<br>
3. Adjust positions of lens and acquire an image of the F object on the screen<br>
4. Measure the height of F image $h_{f}$, distance between screen and the Lens A $d_{SA}$ as well as the distance between Lens D and the F object $d_{DF}$ and the distance between the Lens A and Lens D $d_{AD}$ and record them.<br>
4. Change the distances and collect multiple independent data.<br>

#### Data Collection

In [178]:
pf = pd.read_csv("lensD.csv")
pf

Trial   ddf   dsa   dad  del_ddf  del_dsa  del_dad
0      1   5.0  29.1  53.2     0.05     0.05     0.05
1      2   5.0  41.7  27.0     0.05     0.05     0.05
2      3  14.7  35.4  25.5     0.05     0.05     0.05
3      4  14.7  50.6  19.5     0.05     0.05     0.05

#### Data Analysis

In [179]:
# Calculating focal length of lens D
pf["doa"] = 1/(1/17.6 - 1/pf["dsa"])
pf["did"] = pf["doa"] - pf["dad"]

In [180]:
newval1 = -pf.at[1, "did"] 
pf.at[1, "did"] = newval1
newval2 = -pf.at[2, "did"] 
pf.at[2, "did"] = newval2
newval3 = -pf.at[3,"did"]
pf.at[3,"did"] = newval3
pf

Trial   ddf   dsa   dad  del_ddf  del_dsa  del_dad        doa       did
0      1   5.0  29.1  53.2     0.05     0.05     0.05  44.535652 -8.664348
1      2   5.0  41.7  27.0     0.05     0.05     0.05  30.453112 -3.453112
2      3  14.7  35.4  25.5     0.05     0.05     0.05  35.002247 -9.502247
3      4  14.7  50.6  19.5     0.05     0.05     0.05  26.986667 -7.486667

In [181]:
pf["fd"] = 1/(1/pf["did"] + 1/pf["ddf"])
pf

Trial   ddf   dsa   dad  del_ddf  del_dsa  del_dad        doa       did  \
0      1   5.0  29.1  53.2     0.05     0.05     0.05  44.535652 -8.664348   
1      2   5.0  41.7  27.0     0.05     0.05     0.05  30.453112 -3.453112   
2      3  14.7  35.4  25.5     0.05     0.05     0.05  35.002247 -9.502247   
3      4  14.7  50.6  19.5     0.05     0.05     0.05  26.986667 -7.486667   

          fd  
0  11.822496  
1 -11.161481  
2 -26.873735  
3 -15.257024

In [184]:
f = np.mean(pf["fd"])
f

-10.367435921098462

### Error Propagation

$\delta_{df} = \dfrac{1}{\left(\dfrac{1}{d_{od}} + \dfrac{1}{d_{of}}\right)^{2}}\sqrt{\left(\dfrac{\delta_{od}}{d_{od}}\right)^{2} + \left(\dfrac{\delta_{df}}{d_{df}}\right)^{2}}$   


$\delta_{od} = \dfrac{1}{\left({d_{oa} - d_{ad}}\right)^{2}}\sqrt{\delta_{doa}^{2} + \delta_{dad}^{2}}$

$\delta_{doa} = \dfrac{1}{\left(\dfrac{1}{d_{fa}} + \dfrac{1}{d_{dsa}}\right)^{2}}\sqrt{\left(\dfrac{\delta_{fa}}{d_{fa}}\right)^{2} + \left(\dfrac{\delta_{dsa}}{d_{dsa}}\right)^{2}} = $   


In [ ]:
del_df = 

### Diverging Mirror F

#### Method 1 Lens Clock

__Calibration:__ 

- The dial of the clock is calibrated to display the optical power of a surface made of crown glass (n = 1.53). To read curvatures, therefore, we need to make a correction:<br>
<center>$\dfrac{1}{𝑅} = \dfrac{lensclockreading}{0.53}$</center><br>
- Our lens clock DC offset was precisely calibrated to zero.

__Measure__   
Two independent trials both yielded the result <br>
curvature $C = -1.25\pm0.25 unit$.  


__Analysis__   

$R_{1}=\dfrac{1.25}{0.53}= 2.36\pm0.13$   

The focal length of mirror F is <br><center>$f\approx\frac{R_{1}}{2}=1.18\pm0.07$</center>




#### Method 2 
1. Mount mirror F, F object, and the bench lamp as the diagram below shows.   
2. Place converging lens A and the screen side way, adjust the direction of mirror F so that the reflected ray goes through lens A and form a real image on the screen.
<img src="mirrorF.png" width = "600"/><br><br>  
3. Adjust positions of lens and acquire an image of the F object on the screen<br>
4. Measure the distance between screen and the Lens A $d_{as}$ as well as the distance between mirror F and the F object $d_{mf}$ and the distance between the Lens A and Mirror F $d_{ma}$ and record them.<br>
4. Change the distances to form different sizes of clear real images on the screen and collect multiple independent data.

__Data Collection__




In [147]:
pf = pd.read_csv("mirrorF.csv")
pf

Trial   Dmf   Dma   Das
0      1  26.8  28.3  31.8
1      2  26.8  17.3  45.2
2      3  18.8  24.2  36.6
3      4  18.8  16.8  51.3
4      5  44.6  31.7  28.7
5      6  44.6  58.2  23.5

In [153]:
#calculating focal length of miror F

pf["doa"] = 1/(1/17.6 - 1/pf["Das"])
pf["dim"] = pf["doa"] - pf["Dma"]
pf["fm"] = 1/(1/pf["dim"] + 1/pf["Dmf"])
pf

Trial   Dmf   Dma   Das        doa        dim         fm
0      1  26.8  28.3  31.8  39.414085  11.114085   7.856116
1      2  26.8  17.3  45.2  28.823188  11.523188   8.058344
2      3  18.8  24.2  36.6  33.903158   9.703158   6.399970
3      4  18.8  16.8  51.3  26.791691   9.991691   6.524236
4      5  44.6  31.7  28.7  45.506306  13.806306  10.542719
5      6  44.6  58.2  23.5  70.101695  11.901695   9.394684

In [158]:
fm = np.mean(pf['fm'])
fm

8.129345134327423

In [141]:
-1.5/0.53

-2.830188679245283

In [142]:
finverse = (1.53 - 1)*(1/-2.83 - 1/2.83)
f = 1/finverse
f

-2.6698113207547167

In [33]:
0.53*0.25

0.1325

In [69]:
0.13/(2.83*2.83)

0.01623194196456442

In [70]:
3/0.53

5.660377358490566

In [71]:
fdinverse = 0.53*(1/-5.66 - 1/5.66)
fd= 1/fdinverse
fd

-5.339622641509433

In [72]:
0.34/5.66/5.66

0.01061319282298443

In [146]:
1.25/0.53

2.3584905660377355